src/Glimpse.Server.Web/Storage/InMemoryStorage.cs

```diff
@@ -8,65 +7,200 @@ namespace Glimpse.Server.Web
 {
     public class InMemoryStorage : IStorage, IQueryRequests
     {
+        /// <summary>
+        /// Internal class to contain all data associated with a single request. 
+        /// </summary>
+        private class RequestInfo
+        {
+            public RequestInfo(LinkedListNode<Guid> requestIdNode)
+            {
+                this.RequestIdNode = requestIdNode;
+                this.Messages = new LinkedList<IMessage>();
+                this.Indices = null;
+            }
+
+            /// <summary>
+            /// List of individual messages associated with this request
+            /// </summary>
+            public LinkedList<IMessage> Messages { get; private set; }
+            
+            /// <summary>
+            /// The node in the _activeRequests list.  Storing this here allows us to determine to track 
+            /// requests by age in constant time.
+            /// </summary>
+            public LinkedListNode<Guid> RequestIdNode { get; private set; }
+
+            /// <summary>
+            /// Indices associated with this Request
+            /// </summary>
+            public RequestIndices Indices { get; set; }
+        }
+
         public const int RequestsPerPage = 50;
-        public const int MaxRequests = 500; // TODO: Make this configurable
+        public const int DefaultMaxRequests = 500;
 
-        private readonly List<IMessage> _messages;
-        private readonly ConcurrentDictionary<Guid, RequestIndices> _indices;
+        /// <summary>
+        /// Primary storage for Messages. 
+        /// </summary>
+        private readonly Dictionary<Guid, RequestInfo> _requestTable;
 
-        private static readonly object _locker = new object(); // HACK :(
+        /// <summary>
+        /// Currently active requests, ordered by incoming messages for the request.  The first entry in the list will be
+        /// the most recent request to receive a message, the last entry the oldest request to receive a message.
+        /// </summary>
+        private readonly LinkedList<Guid> _activeRequests;
```

Comment:
These should be generally consistent.  The only time they’ll be inconsistent is during a Persist() call, which should be guarded by a lock.  The various read calls may need to be synchronized to ensure consistent view of the data (e.g., you may be able to read a message that hasn’t yet updated its request’s indices).   The linked-list provides constant time access to track the most recent request & identify the oldest request, with linear additional memory.

I added a CheckConsistency() method to go through and confirm the two structures are consistent.


From: Anthony van der Hoorn [mailto:notifications@github.com]
Sent: Friday, September 25, 2015 10:12 AM
To: Glimpse/Glimpse.Prototype <Glimpse.Prototype@noreply.github.com>
Cc: Mike Kaufman <Mike.Kaufman@microsoft.com>
Subject: Re: [Glimpse.Prototype] 55 in memory storage datastructures (#62)


In src/Glimpse.Server.Web/Storage/InMemoryStorage.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fGlimpse%2fGlimpse.Prototype%2fpull%2f62%23discussion_r40452583&data=01%7c01%7cmike.kaufman%40microsoft.com%7c6aacb86ce43e4c88998d08d2c5cc7bd8%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=lOJmIBEkH0fGpSsA3zdleua%2blsRZ4G6KbE%2bVUuNBnpo%3d>:

>

> -        private static readonly object _locker = new object(); // HACK :(

> +        /// <summary>

> +        /// Currently active requests, ordered by incoming messages for the request.  The first entry in the list will be

> +        /// the most recent request to receive a message, the last entry the oldest request to receive a message.

> +        /// </summary>

> +        private readonly LinkedList<Guid> _activeRequests;

Is the same records indexed in both _activeRequests and _requestTable at the same time? Or can some records be in _activeRequests and not in _requestTable?

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fGlimpse%2fGlimpse.Prototype%2fpull%2f62%2ffiles%23r40452583&data=01%7c01%7cmike.kaufman%40microsoft.com%7c6aacb86ce43e4c88998d08d2c5cc7bd8%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=LmBhCketuU06aW2ISOH3vUzKHm3UVw6iJl0IBcdEpEI%3d>.